Adding extension support

pacostas opened the pull request:
<!-- Thanks for contributing. To speed up the process of reviewing your pull
request please provide us with the following information: -->

## Summary
<!-- A short explanation of the proposed change -->
This PR adds the --extension flag option for packaging an extension.

## Use Cases
With this PR is possible to package an extension by using below command.
```
jam pack --extension  ../path/to/extension.toml  --output ../path/to/build/extension.tgz --version "1.2.3"
```
## Checklist
<!-- Please confirm the following -->
* [ ] I have viewed, signed, and submitted the Contributor License Agreement.
* [ ] I have linked issue(s) that this PR should close using keywords or the Github UI (See [docs](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue))
* [ ] I have added an integration test, if necessary.
* [ ] I have reviewed the [styleguide](https://github.com/paketo-buildpacks/community/blob/main/STYLEGUIDE.md) for guidance on my code quality.
* [ ] I'm happy with the commit history on this PR (I have rebased/squashed as needed).

robdimsdale commented:
Thanks for submitting this @pacostas 

We'll take a look at trying out this functionality and give you feedback on it.

pacostas commented:
@robdimsdale I've added it as a draft, as the tests are not implemented yet and also I'm not sure if this heads to the right direction.
Probably next week I'll be able to sign the CLA.
Thank you :)

sophiewigmore commented:
@pacostas to me, this PR seems like its definitely heading in the right direction!

sophiewigmore commented:
@pacostas to me, this PR seems like its definitely heading in the right direction!

mhdawson commented:
Successfully used this in an integration test in https://github.com/nodeshift/ubi-nodejs-extension/pull/35

sophiewigmore commented:
@pacostas let us know when this PR is officially ready for review or if you need feedback on anything particular

pacostas commented:
@sophiewigmore I dont have anything in particular to ask, so I think is ready for review :)

sophiewigmore commented on commands/pack.go:
```diff
@@ -56,24 +56,33 @@ func init() {
 
 	configParser := cargo.NewBuildpackParser()
-	config, err := configParser.Parse(flags.buildpackTOMLPath)
+	config, err := configParser.Parse(buildpackOrExtensionTOMLPath)
```
Should we hold off on this PR until the extension parser is available in packit? (https://github.com/paketo-buildpacks/packit/pull/489)

mhdawson replied:
I had suggested @pacostas not block this PR on the one in packit in order to avoid a long thread of "this waits for that etc." Having said that, as @pacostas said if https://github.com/paketo-buildpacks/packit/pull/489 won't take too long to land, then waiting and updating this PR makes sense.

sophiewigmore replied:
That PR is pretty reasonable, I don't think it'll be too controversial most likely! Unfortunately, this type of packit-jam interdependency is pretty common

robdimsdale replied:
That other PR looks fine. I'd suggest holding off on this PR until that lands (and packit is released), and then reworking this PR to match that. I think it'll make this code cleaner.

pacostas replied:
I'm glad that the PR on packit has been merged!! working on using it on this one :)

sophiewigmore replied:
I have released your changes in https://github.com/paketo-buildpacks/packit/releases/tag/v2.10.0 so you can use it here!

sophiewigmore commented:
Even though the cases are very similar, I'd love to see an extra test case added for the `extension.toml` case

robdimsdale commented on commands/pack.go:
```diff
@@ -56,24 +56,33 @@ func init() {
-	defer os.RemoveAll(buildpackDir)
+	defer os.RemoveAll(buildpackOrExtensionDir)
+
+	var buildpackOrExtensionTOMLPath string
```
Similar to above, can we name this something shorter like `configPath`?

pacostas replied:
sure :)

pacostas commented on commands/pack.go:
```diff
@@ -56,24 +56,33 @@ func init() {
 
 func packRun(flags packFlags) error {
-	buildpackDir, err := os.MkdirTemp("", "dup-dest")
+	buildpackOrExtensionDir, err := os.MkdirTemp("", "dup-dest")
```
packageDir sounds more descriptive

pacostas replied:
After creating a separate method for handling packing the extension, on my latest commit https://github.com/paketo-buildpacks/jam/pull/234/commits/563b76266b810a49a7516f1812858f94b53611ff I reverted the changes on the packRun function and renamed only the `buildpackdir` to `tmpDir`

pacostas commented:
The final output, looks like it will end up with a lot of if statements handling the scenario "if is buildpack or extension" each time the buildpack and extension differ, something like:
```
if isBuildpack{
	config.Buildpack.Version = flags.version
} else {
config.Extension.Versions = flags.version
}
```
  
I think it would be better to create a separate method that will handle only the extensions. The only drawback with that is that it ends up with a few lines of duplicate code which compared to the multiple "if statements" solution seems more manageable.
I'm sharing in case you have any other suggestions you would like to see,  in terms of a more maintainable solution.

pacostas commented:
> @pacostas
> 
> > I think it would be better to create a separate method that will handle only the extensions. The only drawback with that is that it ends up with a few lines of duplicate code which compared to the multiple "if statements" solution seems more manageable.
> 
> I think this is fine. A few lines of duplicated code sounds potentially better than lots of `if` statements. We will be able to evaluate for sure once we see it, so I would suggest trying that option out and pushing it to this branch.

Below commits is the implementation using a separate function.
https://github.com/paketo-buildpacks/jam/pull/234/commits/f808f6ff0911965e5866128bfd2f9c17b50655e3
https://github.com/paketo-buildpacks/jam/pull/234/commits/6abd3ed7e551f20a554f89df1c58d228bef99b6c
https://github.com/paketo-buildpacks/jam/pull/234/commits/563b76266b810a49a7516f1812858f94b53611ff

What is missing is that I havent added tests yet for packing the extension.
Also the `HasStack` method is missing, I will open soon another PR on packit to add the `HasStack` method, the tests will fail till then.

pacostas commented:
This PR has to be merged after https://github.com/paketo-buildpacks/packit/pull/490

mhdawson commented:
For anybody looking https://github.com/paketo-buildpacks/packit/pull/490 has already landed :)

sophiewigmore commented:
@pacostas @mhdawson we will take another look this week! thanks for the updates

pacostas commented:
> @pacostas @mhdawson we will take another look this week! thanks for the updates
@sophiewigmore 

I havent finish it yet, as I havent finished with unit/integration tests. I'll submit them soon almost 90% done

pacostas commented:
@sophiewigmore I've finished with this PR Implementation, is ready for review. Although this PR wont pass the tests till this PR https://github.com/paketo-buildpacks/packit/pull/490 has being included on the latest release version of https://github.com/paketo-buildpacks/packit due to a recent change that I had to add.

robdimsdale commented:
I released [packit v2.10.1](https://github.com/paketo-buildpacks/packit/releases/tag/v2.10.1) which includes the `HasStack` method. You should be good to update to that and fix the tests.

Also, I think you can refactor the `dependency_cacher.go` file using an unexported common method and an unexported interface. See [this commit](https://github.com/paketo-buildpacks/packit/blob/1ad99b1419b606ac2a019a872a074709a66e99b3/scribe/emitter.go) for inspiration. I think this will make the code much easier to read, as there is a fair amount of duplication across the `Cache`/`CacheExtension` methods. Because this would be internal/unexported interfaces and functions, it doesn't change the exported interface and so nothing would change about the tests or the consumption of the library.

sophiewigmore commented:
+ 1 to Rob's comment, and I think we can do the same refactor on file_bundler as well

pacostas commented:
> I released [packit v2.10.1](https://github.com/paketo-buildpacks/packit/releases/tag/v2.10.1) which includes the `HasStack` method. You should be good to update to that and fix the tests.
> 
> Also, I think you can refactor the `dependency_cacher.go` file using an unexported common method and an unexported interface. See [this commit](https://github.com/paketo-buildpacks/packit/blob/1ad99b1419b606ac2a019a872a074709a66e99b3/scribe/emitter.go) for inspiration. I think this will make the code much easier to read, as there is a fair amount of duplication across the `Cache`/`CacheExtension` methods. Because this would be internal/unexported interfaces and functions, it doesn't change the exported interface and so nothing would change about the tests or the consumption of the library.

I suppose that the Cache method on `dependency_cacher` is similar to the [launch](https://github.com/paketo-buildpacks/packit/blob/1ad99b1419b606ac2a019a872a074709a66e99b3/scribe/emitter.go#LL174C13-L174C13) method where the launch method calls the same method e.x. [GetType](https://github.com/paketo-buildpacks/packit/blob/1ad99b1419b606ac2a019a872a074709a66e99b3/scribe/emitter.go#L182) for [launchProcess]( https://github.com/paketo-buildpacks/packit/blob/1ad99b1419b606ac2a019a872a074709a66e99b3/scribe/emitter.go#L14) inteface . 

So from what I see that this could work  in our case, is creating an interface on the `dependency_cacher.go` 
```
type ConfigBuilpackOrExtensionMetadataDependency interface {
	GetChecksum() string
}
```
Changing the declaration of the Cache method to accept that interface instead of the `[]cargo.ConfigExtensionMetadataDependency` or `[]cargo.ConfigMetadataDependency`
```
func (dc DependencyCacher) Cache(root string, deps []ConfigBuilpackOrExtensionMetadataDependency) ([]cargo.ConfigMetadataDependency, error) {

```
and each time we need a a value from the fields of the struct we need to access them with a method like so 
```
		checksum := dep.GetChecksum()
		_, hash, _ := strings.Cut(dep.GetChecksum(), ":")
```

and also having that method implemented on both `ConfigMetadataDependency` and `ConfigExtensionMetadataDependency` struct like so 
```
func (cd ConfigExtensionMetadataDependency) GetChecksum() string {
	return cd.Checksum
}
```

```
func (cd ConfigMetadataDependency) GetChecksum() string {
	return cd.Checksum
}
```

Although I havent figured out how the return type can be implemented without breaking the interface of the cache function as it will return a subset of the `ConfigExtensionMetadataDependency` or `ConfigMetadataDependency` structs as the `ConfigBuilpackOrExtensionMetadataDependency` will always be a subset of these two. The [launch](https://github.com/paketo-buildpacks/packit/blob/1ad99b1419b606ac2a019a872a074709a66e99b3/scribe/emitter.go#LL174C13-L174C13) method doesnt return anything and might that be the reason why this kind of implementation is suitable. 

I'm new to Go language so I'm sharing cause probably there is something that I'm not aware of, as I dont have enough context of what Go language is capable of in that case.

Also this implementation seems like will produce more code from what it is now, as for each field we need to create a new method, although the good thing is that there will be no duplicate code on the main logic which is good as the `ConfigMetadataDependency` and `ConfigExtensionMetadataDependency`  probably will remain the same when it comes to caching.

mhdawson commented:
I had a chat with @pacostas on the suggested approach a few days ago and I think he's working on an update to the PR to refect that. @pacostas let me know if I have that wrong.

pacostas commented:
@sophiewigmore I didnt inform because I'm still unsure if I'll manage to make this implementation work :)  so yes, that's correct, I'm working on an update @mhdawson

pacostas commented:
@robdimsdale I've finished with the implementation of `cache` and `cacheExtension` in similar pattern as of the [example](https://github.com/paketo-buildpacks/packit/blob/1ad99b1419b606ac2a019a872a074709a66e99b3/scribe/emitter.go) you sent me on this [commit](https://github.com/paketo-buildpacks/jam/pull/234/commits/1bca772bec717e9906207d922339b5295a9069ed) . If the commit looks good, I can also proceed in refactoring `fileBunlder` in a similar way.

robdimsdale commented on commands/pack.go:
```diff
@@ -56,24 +56,37 @@ func init() {
+	config, err := configParser.Parse(buildpackTOMLPath)
 	if err != nil {
-		return fmt.Errorf("failed to parse buildpack.toml: %s", err)
+		return fmt.Errorf("failed to parse .toml file : %s", err)
```
It might be good to a print the file name here instead of just `.toml file`.

E.g.:

```
failed to parse buildpack.toml
```

or:

```
failed to parse extension.toml
```

robdimsdale commented on internal/dependency_cacher.go:
```diff
@@ -28,33 +28,98 @@ func NewDependencyCacher(downloader Downloader, logger scribe.Logger) Dependency
 }
 
-func (dc DependencyCacher) Cache(root string, deps []cargo.ConfigMetadataDependency) ([]cargo.ConfigMetadataDependency, error) {
+type ConfigBuilpackOrExtensionMetadataDependency interface {
```
I think this interface should be unexported (i.e. `type configBuilpackOrExtensionMetadataDependency interface`)

robdimsdale commented on internal/dependency_cacher.go:
```diff
@@ -28,33 +28,98 @@ func NewDependencyCacher(downloader Downloader, logger scribe.Logger) Dependency
+	GetVersion() string
+}
+
+type ExtensionConfigMetadataDependency struct {
```
Similarly, this should be unexported (i.e. `extensionConfigMetadataDependency`)

robdimsdale commented on internal/dependency_cacher.go:
```diff
@@ -28,33 +28,98 @@ func NewDependencyCacher(downloader Downloader, logger scribe.Logger) Dependency
+	return cd.Version
+}
+
+type BuildpackConfigMetadataDependency struct {
```
And this should be unexported (i.e. `buildpackConfigMetadataDependency`)

robdimsdale reviewed (commented):
I think this is almost there. There's a couple of types we can make unexported, which keeps them out of the public interface of the package, and there's another minor logging change I think will improve UX. All comments inline.

pacostas commented:
@robdimsdale I've added your suggestions.
* https://github.com/paketo-buildpacks/jam/pull/234/commits/532c3271c762bbadffe4753dbbb7bf6f2f4bccd9
* https://github.com/paketo-buildpacks/jam/pull/234/commits/1e691b2f09be5feba45f5040b3b8161e395b20c3

 I've also refactored Bunlde and ExtensionBundle for avoiding duplicate code as @sophiewigmore  has suggested
* https://github.com/paketo-buildpacks/jam/pull/234/commits/946a8d1a418c9e950bf93ae98ad878478a64299f

sophiewigmore commented:
I'll take another look in the next few days, sorry about the delay

sophiewigmore commented:
@pacostas these changes look great. Thanks for the hard work on this PR 🔥

sophiewigmore commented:
I cannot merge it in due to conflicts, if you can resolve those
